### ISSUE TITLE
Pdf annotations not migrated properly

### ISSUE THREAD

**Issue by ichnilatis-gr** [koreader/koreader#11762]
* KOReader version: v2024.04-9
* Device: PB 628

#### Issue
After having installed the nightly version v2024.04-9, some pdf files were not recognised and the message "No reader engine for this file or invalid file" was displayed. It didn't happened with all the pdf files and I can't see what was special about these files that caused this problem.

**Comment by Frenzie** [comment]
The crash.log will say something like `doShowReader coroutine crashed`. Please share that part of the log and/or the actual file.

**Comment by Frenzie** [comment]
That file works for me. Could you describe a bit more what exactly you're doing? Maybe it's the same as #11760.

**Comment by 672493** [comment]
I could successfully open it after removing sdr folder. So the problem is probably with metadata.pdf.lua.
[metadata.pdf.lua.zip](https://github.com/koreader/koreader/files/15220852/metadata.pdf.lua.zip)
Sorry, I can't upload Lua file.

**Comment by 672493** [comment]
Anyway, file is opened correctly, and it's not a problem)

**Comment by hius07** [comment]
Coverting higlights to annotations. I'm lookin into it.

**Comment by hius07** [comment]
The bug in converting old **page bookmarks** in **pdf** files. Is being fixed today.

**Comment by ichnilatis-gr** [comment]
> The crash.log will say something like `doShowReader coroutine crashed`. Please share that part of the log and/or the actual file.

Let me mention that the problem doesn't exist in the version v2024.04, which I have installed again. I think there is no need for me to look for this part of the log any longer. Don't I?

**Comment by ichnilatis-gr** [comment]
This fixed the problem. Thank you.

**Comment by Frenzie** [comment]
> I think there is no need for me to look for this part of the log any longer. Don't I?

Indeed, that'll be a bug in #11563. Thanks for testing!

**Comment by ichnilatis-gr** [comment]
Since the title of the issue has changed, I think it's here that I have to mention some problems with the notes. 
I installed the nightly version v2024.04-15 yesterday and I added the readerannotation.lua provided above. Then I noticed that, when I added a note to a highlighted text, the note mark didn't appear and the note text wasn't displayed in the bookmark list, where, after long press on a highlight, there was the option "Add note" instead of "Edit note". The only way to edit the note was to tap on on the highlighted text and then on the option "Edit highlight" and "Edit note".

In the epub books, when I added a note to a highlight, it was displayed in the bookmark list, but if I long pressed the highlight and then selected the option "Edit note", the note text was not displayed in the pop up window. In this case I had to follow the same procedure as above, in order to edit or delete the note.

**Comment by hius07** [comment]
Does it happen on any file?
Please describe the steps to reproduce the issue.

**Comment by ichnilatis-gr** [comment]
It's a little bit strange. Now I notice these: 
In a pdf file, If I make a highlight and then I add a note to it, there is the above mentioned problem. If I select a text and then I select the option "Add note" in the pop-up menu, the note mark appears and the note is displayed in the bookmarks list. But if I long press the highlight in the bookmarks list and select "Edit note", the note is not displayed in the pop-up window. 
If I tap on a highlight (with a note) in the text and select "Delete", the highlight is not deleted. Also, if I tap on the highlight, there is no option "Edit note" but "Add note", even if there is a note added to the highlight. I can delete the highlight (and the note) only after I long press the highlight in the bookmarks list and then select "Delete".

**Comment by hius07** [comment]
Sorry, cannot reproduce with my pdf files.
Can you make one problematic note and show me the metadata file (only, no book).

**Comment by hius07** [comment]
PDF page bookmarks conversion fixed in https://github.com/koreader/koreader/pull/11761.

**Comment by ichnilatis-gr** [comment]
> Sorry, cannot reproduce with my pdf files. Can you make one problematic note and show me the metadata file (only, no book).

Am I late again! :-)
Do you still want me to upload the metadata file? I fount out that the problems occur in books that have already highlights. I also noticed that in the case, when I first highlight the text and then add a note to it, the note is attached to the first highlight of the book and thus it is not displayed in the bookmarks list; it appears in the first highlight.

**Comment by hius07** [comment]
> Do you still want me to upload the metadata file?

Yes please.

**Comment by ichnilatis-gr** [comment]
I'm sorry, I don't see that the bugs have been fixed.

**Comment by hius07** [comment]
As usual, did you remove `txt` extension and put it to koreader/frontend/apps/reader/modules, then restart?

**Comment by ichnilatis-gr** [comment]
Of course.
Now, if I highlight a text and then tap on it, the note "note1" or another that I have already done appears in the pop-up window. If I tap on the highlight and select "Delete" it isn't deleted.
If I select a text and then the option "Add note" and make a note, the note mark appears, but if I try to delete the highlight by taping on it, it isn't deleted.

**Comment by hius07** [comment]
Cannot reproduce, sorry.
Is it a new issue, "Delete" button doesn't delete the highlight?
For all highlights or for highlights with notes only? "Delete highlight" button in the dialog with a note or "Delete" button in the highlight popup dialog?

The above module should fix adding a note to wrong highlight from the bookmark list with reverse sorting order.

**Comment by ichnilatis-gr** [comment]
I'm talking about the "Delete" button, after tapping on a highlight, whether it has a note or not. If I highlight a text and then tap on it, the note window opens with a note text that I have already added to another highlight.
I can delete the highlights only via the relative button, after long pressing a highlight in the bookmarks list.

Also new notes continue to be added to wrong highlight.

**Comment by ichnilatis-gr** [comment]
The bugs seem to have been solved. Thank you!

Let me just mention one thing I noticed. I can't remember if it existed before. When we open the "bookmark details" window via the bookmarks list and then select "Edit note" or "Edit text", after editing we don't return to the "bookmark details" window but to the bookmarks list. Is it perhaps a little strange to go two steps behind?

**Comment by hius07** [comment]
Thank you for testing.
Yes, this behaviour didn't change, we close Bookmark details on any operation.

**Comment by hius07** [comment]
> When we open the "bookmark details" window via the bookmarks list and then select "Edit note" or "Edit text", after editing we don't return to the "bookmark details" window but to the bookmarks list. Is it perhaps a little strange to go two steps behind?

Changed in https://github.com/koreader/koreader/pull/11825.
